Add LDAP->Django Group mapping

alandmoore opened the pull request:
This PR addresses issue #11 by adding a GROUP_MAP setting which allows an LDAP group to be mapped to a list of Django groups. The listed groups will be added or removed from a user's account at login time based on the users membership or non-membership in the LDAP group.

alandmoore committed:
Added GROUP_MAPPING feature

Added a GROUP_MAPPING feature, where LDAP groups can be mapped to a list
of Django groups.  Django groups are added to or removed from the user
based on membership or non-membership (respectively) in the LDAP group.

The setting AUTH_LDAP_GROUP_MAP was added which defines this.

alandmoore committed:
Added myself to authors

alandmoore committed:
Remove groups first

Remove groups first, so that group additions take priority.

alandmoore committed:
Merge branch 'master' of https://github.com/alandmoore/django_auth_ldap3

sjkingo commented:
Thanks Alan, this looks good.

Remembering how I implemented the LDAP group support, I wonder if we could somehow simplify how it works, instead of patching on some related functionality. Presently there is support to restrict login by LDAP group (`AUTH_LDAP_LOGIN_GROUP`, `AUTH_LDAP_ADMIN_GROUP`). Perhaps we could merge that code with your patch, so that we can sync the LDAP group with Django, and use that for authorization?

I haven't had much time to sit down and hash it out so I'm not sure if that would work or not.

What are your thoughts?

alandmoore commented:
I think it would be better to leave them separate.  Even though they all deal with LDAP groups, on the Django side they map to conceptually different things.  
It might be nice to have a setting to map a group to superusers, but that's another ball of wax.

alandmoore committed:
Use get_user_model to determine the user model

Instead of directly importing User from django.contrib.auth.models, use
the get_user_model function to determine the configured user model.

This fixes a failure when trying to authenticate in an application that
uses a custom user model.

sjkingo commented:
Could you possibly rebase this PR to ensure it works properly with your latest change, and then I can merge this too and release a new version on PyPi.

alandmoore commented:
Let me just close this and open a fresh one.
